### ISSUE TITLE
Area/Position metatables/classes

### ISSUE THREAD

**Issue by Nexela** [Afforess/Factorio-Stdlib#97]
Choumiko commented 5 hours ago
[x] position1 + position2 -- modifies position1 by adding position2 to it
[x] position1:add(position2) -- same as above
[x] local position3 = position1:add(position2) -- same as above with position3 being a reference to position1
Not sure i like the idea of an addition changing one of the operands, if i understand correctly each of these operations would mutate position1 ? To me addition/subtraction are pure mathematical/functional and therefore shouldn't mess with the operands?
Also (at least for the operator overwriting) it would make addition non commutative, since
position1 + position2 -- {1,1} + {2,2}
position1 + position3 -- {3,3} + {0,1} = {3,4}
would lead to a different result than
position2 + position1 -- {2,2} + {1,1}
position1 + position3 -- {1,1} + {0,1} = {1,2}

Hope that makes sense, but as long as Position.add(pos1, pos2) still work as it does now i'm fine with it. Just recently started changing FARL to use Position, so i don't wanna touch it again too soon ;)

**Comment by Nexela** [comment]
In order to maintain most backwards compatibility I have made some sacrifices.
pos1 + pos2 actually calls Position.add(pos1, pos2)

The only thing not backwards compatible is a "new" position is no longer returned instead pos1 is mutated, 
entity.position  returns a "new" position table,   writing to this table doesn't update the entities position.

But I can see where pos1 + pos2 could cause some issues so I will change it so the operand functions return a new position without mutating the original.

pos1 = pos1 + pos2  -- return a new position and store it in pos1

As for the rest mutating position would only need to be changed in cases like this:

local pos = entity.position
pos:expand_to_area(4) -- pos is now a bounding_box
area = pos:expand_to_area(4) -- pos is now a bounding box, area is a refrence to pos

Actually looking on this it makes sense for expand_to_area to return new also so that is a bad example :)

**Comment by Nexela** [comment]
@Choumiko I have pushed some of the work to https://github.com/Nexela/Factorio-Stdlib/tree/area-position-metatables If you want to take a peek at it.

Right now the only drawback I need to figure out is how to make it more on_load friendly if using them as classes.

**Comment by Choumiko** [comment]
> entity.position returns a "new" position table, writing to this table doesn't update the entities position.

In FARL i use a few lookup tables for offsets and also some fake entities (a lua table with name, position etc) which could be affected.

If i don't miss anything/my knowledge of metatables doesn't fail me, it always seems to use Position.new() on positions before doing stuff with them? That would solve all my concerns then since no operand is being mutated?

After writing the first comment yesterday i realized that i don't have to update FARLs stdlib if i don't want to, so my objections are mostly cleared :D

> Right now the only drawback I need to figure out is how to make it more on_load friendly if using them as classes.

In FARL i do it like this: https://github.com/Choumiko/FARL/blob/master/control.lua#L18
Since there's no way of knowing if a position gets stored in global and where, that's probably for the mod dev to worry about.
You could probably add some sort of helper function, that get's passed some table keys and checks those for tables containing x and y as a key and  then set the metatable?

**Comment by Nexela** [comment]
> If i don't miss anything/my knowledge of metatables doesn't fail me, it always seems to use Position.new() on positions before doing stuff with them? That would solve all my concerns then since no operand is being mutated?

If I did everything right (and I think I did! as all tests passed without being changed)
Position.new is called pretty much every time Position is used except new is probably a bad name since it isn't always new.

First checks if Position is not nil, If it has a meta table return itself, if it doesn't have a metatable return a new Position table correctly formatted with metatable

+/- Position.add  Position.subtract return New positions since yeah mutating for maths would be bad
a = Position.new({1, 2})
b =Position.new({1, 3})
c = Position.add(a, b)  returns {2, 5}  without affecting a or b
c = a + b same as above

On the flip side
Position.translate(a, 0)  does mutate a and would return a as {1, 1}
This allows
Position.translate(a, 0)
a = Position.translate(a, 0)
a:translate(0)
a = a:translate(0)
b = a:translate(0)
To all have the same results. 

TL:DR, 95% backwards compatabilty, The only times it would be an issue is if modders are doing
b = Position.translate(a, 0)
c = Position.translate(a, 4)
And expecting a to not change  If that is the case then doing this should fix em up
b = Position.translate(Position.copy(a), 0)

> Since there's no way of knowing if a position gets stored in global and where, that's probably for the mod dev to worry about.

Agreed, I will make a simple helper, similar to what I did with Queue

**Comment by Nexela** [comment]
Closed with 536e9a7

**Comment by Afforess** [comment]
@Nexela you might also want to check the wiki page for stdlib and update any examples to use the new metatables where appropriate.

**Comment by Nexela** [comment]
The metatables are completely optional. In fact if you don't know about them you won't notice any change
(almost)  by default area tables are mutated in place.

However a quick primer on their use and benefits would be nice, Adding that the 1.0 release todo
